File: src/calcio_manager/models/player.py
"""Player data models."""

from uuid import UUID, uuid4

from pydantic import BaseModel, Field

from calcio_manager.models.enums import MoraleLevel, PlayerRole


class TechnicalAttributes(BaseModel):
    """Technical skills of a player (1-20 scale)."""

    passing: int = Field(ge=1, le=20, description="Passaggio")
    dribbling: int = Field(ge=1, le=20, description="Dribbling")
    finishing: int = Field(ge=1, le=20, description="Tiro")
    first_touch: int = Field(ge=1, le=20, description="Controllo palla")


class MentalAttributes(BaseModel):
    """Mental attributes of a player (1-20 scale)."""

    positioning: int = Field(ge=1, le=20, description="Posizionamento")
    decisions: int = Field(ge=1, le=20, description="Decisioni")
    leadership: int = Field(ge=1, le=20, description="Leadership")
    teamwork: int = Field(ge=1, le=20, description="Gioco di squadra")


class PhysicalAttributes(BaseModel):
    """Physical attributes of a player (1-20 scale)."""

    pace: int = Field(ge=1, le=20, description="Velocità")
    stamina: int = Field(ge=1, le=20, description="Resistenza")
    strength: int = Field(ge=1, le=20, description="Forza")


class GoalkeepingAttributes(BaseModel):
    """Goalkeeping-specific attributes (1-20 scale)."""

    reflexes: int = Field(ge=1, le=20, description="Riflessi")
    handling: int = Field(ge=1, le=20, description="Presa")


class PlayerAttributes(BaseModel):
    """Complete set of player attributes."""

    technical: TechnicalAttributes
    mental: MentalAttributes
    physical: PhysicalAttributes
    goalkeeping: GoalkeepingAttributes = GoalkeepingAttributes(reflexes=5, handling=5)


class Injury(BaseModel):
    """Player injury information."""

    description: str
    days_remaining: int = Field(ge=0)


class SeasonStats(BaseModel):
    """Player statistics for a single season."""

    season: str  # e.g. "2025-2026"
    appearances: int = 0
    goals: int = 0
    assists: int = 0
    yellow_cards: int = 0
    red_cards: int = 0
    minutes_played: int = 0
    average_rating: float = 0.0


class Player(BaseModel):
    """A football player in the game."""

    id: UUID = Field(default_factory=uuid4)
    first_name: str
    last_name: str
    age: int = Field(ge=14, le=60)
    role: PlayerRole
    overall: int = Field(ge=1, le=20, description="Overall rating derived from attributes")
    attributes: PlayerAttributes
    condition: float = Field(ge=0.0, le=1.0, default=1.0, description="Physical condition 0-1")
    morale: MoraleLevel = MoraleLevel.NORMAL
    injury: Injury | None = None
    suspended: bool = False
    yellow_card_accumulation: int = 0
    current_season_stats: SeasonStats = SeasonStats(season="2025-2026")
    history: list[SeasonStats] = Field(default_factory=list)

    @property
    def full_name(self) -> str:
        """Return the player's full name."""
        return f"{self.first_name} {self.last_name}"

    @property
    def short_name(self) -> str:
        """Return abbreviated name (e.g., 'M. Rossi')."""
        return f"{self.first_name[0]}. {self.last_name}"

    @property
    def is_available(self) -> bool:
        """Check if the player is available for selection."""
        return self.injury is None and not self.suspended

    def calculate_overall(self) -> int:
        """Calculate overall rating based on role and attributes."""
        attrs = self.attributes
        tech = attrs.technical
        ment = attrs.mental
        phys = attrs.physical

        if self.role == PlayerRole.GK:
            gk = attrs.goalkeeping
            score = (
                gk.reflexes * 3 + gk.handling * 3
                + ment.positioning * 2 + phys.pace * 1 + phys.strength * 1
            ) / 10
        elif self.role == PlayerRole.DEF:
            score = (
                ment.positioning * 3 + phys.strength * 2 + phys.pace * 1
                + tech.passing * 2 + ment.teamwork * 1 + ment.decisions * 1
            ) / 10
        elif self.role == PlayerRole.MID:
            score = (
                tech.passing * 3 + ment.decisions * 2 + ment.teamwork * 2
                + phys.stamina * 1 + tech.dribbling * 1 + ment.positioning * 1
            ) / 10
        else:  # FWD
            score = (
                tech.finishing * 3 + tech.dribbling * 2 + phys.pace * 2
                + tech.first_touch * 1 + ment.positioning * 1 + ment.decisions * 1
            ) / 10

        return max(1, min(20, round(score)))

File: src-tauri/src/domain/models/player.rs
use serde::{Deserialize, Serialize};
use uuid::Uuid;

use super::enums::{MoraleLevel, PlayerRole};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TechnicalAttributes {
    pub passing: u8,
    pub dribbling: u8,
    pub finishing: u8,
    pub first_touch: u8,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MentalAttributes {
    pub positioning: u8,
    pub decisions: u8,
    pub leadership: u8,
    pub teamwork: u8,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PhysicalAttributes {
    pub pace: u8,
    pub stamina: u8,
    pub strength: u8,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct GoalkeepingAttributes {
    pub reflexes: u8,
    pub handling: u8,
}

impl Default for GoalkeepingAttributes {
    fn default() -> Self {
        Self {
            reflexes: 5,
            handling: 5,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PlayerAttributes {
    pub technical: TechnicalAttributes,
    pub mental: MentalAttributes,
    pub physical: PhysicalAttributes,
    pub goalkeeping: GoalkeepingAttributes,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Injury {
    pub description: String,
    pub days_remaining: u32,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SeasonStats {
    pub season: String,
    pub appearances: u32,
    pub goals: u32,
    pub assists: u32,
    pub yellow_cards: u32,
    pub red_cards: u32,
    pub minutes_played: u32,
    pub average_rating: f32,
}

impl SeasonStats {
    pub fn new(season: impl Into<String>) -> Self {
        Self {
            season: season.into(),
            appearances: 0,
            goals: 0,
            assists: 0,
            yellow_cards: 0,
            red_cards: 0,
            minutes_played: 0,
            average_rating: 0.0,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Player {
    pub id: Uuid,
    pub first_name: String,
    pub last_name: String,
    pub age: u8,
    pub role: PlayerRole,
    pub overall: u8,
    pub attributes: PlayerAttributes,
    pub condition: f32,
    pub morale: MoraleLevel,
    pub injury: Option<Injury>,
    pub suspended: bool,
    pub yellow_card_accumulation: u32,
    pub current_season_stats: SeasonStats,
    pub history: Vec<SeasonStats>,
}

impl Player {
    pub fn full_name(&self) -> String {
        format!("{} {}", self.first_name, self.last_name)
    }

    pub fn short_name(&self) -> String {
        let initial = self.first_name.chars().next().unwrap_or('?');
        format!("{}. {}", initial, self.last_name)
    }

    pub fn is_available(&self) -> bool {
        self.injury.is_none() && !self.suspended
    }

    pub fn calculate_overall(&self) -> u8 {
        let attrs = &self.attributes;
        let tech = &attrs.technical;
        let ment = &attrs.mental;
        let phys = &attrs.physical;

        let score = match self.role {
            PlayerRole::Gk => {
                let gk = &attrs.goalkeeping;
                (gk.reflexes as f32 * 3.0
                    + gk.handling as f32 * 3.0
                    + ment.positioning as f32 * 2.0
                    + phys.pace as f32
                    + phys.strength as f32)
                    / 10.0
            }
            PlayerRole::Def => {
                (ment.positioning as f32 * 3.0
                    + phys.strength as f32 * 2.0
                    + phys.pace as f32
                    + tech.passing as f32 * 2.0
                    + ment.teamwork as f32
                    + ment.decisions as f32)
                    / 10.0
            }
            PlayerRole::Mid => {
                (tech.passing as f32 * 3.0
                    + ment.decisions as f32 * 2.0
                    + ment.teamwork as f32 * 2.0
                    + phys.stamina as f32
                    + tech.dribbling as f32
                    + ment.positioning as f32)
                    / 10.0
            }
            PlayerRole::Fwd => {
                (tech.finishing as f32 * 3.0
                    + tech.dribbling as f32 * 2.0
                    + phys.pace as f32 * 2.0
                    + tech.first_touch as f32
                    + ment.positioning as f32
                    + ment.decisions as f32)
                    / 10.0
            }
        };

        score.round().clamp(1.0, 20.0) as u8
    }
}
